defuse/src/tokens/nep141/withdraw.rs

```diff
@@ -0,0 +1,112 @@
+use defuse_contracts::{
+    defuse::{
+        tokens::{
+            nep141::{FungibleTokenWithdrawResolver, FungibleTokenWithdrawer},
+            TokenId,
+        },
+        Result,
+    },
+    utils::cache::{CURRENT_ACCOUNT_ID, PREDECESSOR_ACCOUNT_ID},
+};
+use near_contract_standards::fungible_token::core::ext_ft_core;
+use near_sdk::{
+    assert_one_yocto, env, json_types::U128, near, require, serde_json, AccountId, NearToken,
+    PromiseOrValue, PromiseResult,
+};
+
+use crate::{DefuseImpl, DefuseImplExt};
+
+#[near]
+impl FungibleTokenWithdrawer for DefuseImpl {
+    #[payable]
+    fn ft_withdraw(
+        &mut self,
+        token: AccountId,
+        receiver_id: AccountId,
+        amount: U128,
+        memo: Option<String>,
+        msg: Option<String>,
+    ) -> PromiseOrValue<U128> {
+        assert_one_yocto();
+        self.internal_ft_withdraw(
+            PREDECESSOR_ACCOUNT_ID.clone(),
+            receiver_id,
+            token,
+            amount,
+            memo,
+            msg,
+        )
+        .unwrap()
+    }
+}
+
+impl DefuseImpl {
+    fn internal_ft_withdraw(
+        &mut self,
+        sender_id: AccountId,
+        receiver_id: AccountId,
+        token: AccountId,
+        amount: U128,
+        memo: Option<String>,
+        msg: Option<String>,
+    ) -> Result<PromiseOrValue<U128>> {
+        require!(amount.0 > 0, "zero amount");
+
+        self.internal_withdraw(&sender_id, [(TokenId::Nep141(token.clone()), amount.0)])?;
```

Comment:
What if the `ft_transfer_call` below will fail? Isn't it better to move the line to the `ft_resolve_withdraw`?

Reply:
Yes. I see that you make a refund in the `ft_resolve_withdraw`. I guess you need to provide static gas for the `ft_transfer_call` and `ft_resolve_withdraw` to prevent the potential fail in the `ft_resolve_withdraw`.

Reply:
See https://github.com/defuse-protocol/defuse-contracts/pull/15#discussion_r1781792833